sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc

```diff
@@ -0,0 +1,661 @@
+
+## Introduction
+This document presents an ongoing work towards defining a unified matrix interface. This interface is intended to unify different tensor hardware: Intel AMX in CPUs, DPAS in Intel GPUs, Habana Gaudi and Goya tensor and gemm cores, Nvidia TPUs, IBM Power MMA. All these hardware provide low-level intrinsics or assembly to access and perform matrix operations. The goal is to provide a unified interface that is portable but also benefit from the maximum performance these different hardware can offer.
+
+## Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification section 6.3.3 "Feature test macros".  Therefore, an
+implementation supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_MATRIX` to one of the values defined in the table below.
+Applications can test for the existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's APIs the implementation supports.
+
+[frame="none",options="header"]
+|======================
+|Value |Description
+|1     |The APIs of this experimental extension are not versioned, so the feature-test macro always has this value. 
+|======================
+
+## Matrix API Versions
+
+While this document presents the core API that unifies Intel AMX, DPAS, and Nvidia Tensor Cores, the implementations support slightly different versions of the API. For this reason, we introduce a new macro, namely `SYCL_EXT_ONEAPI_MATRIX_VERSION`  to distinguish between these different implementations. The goal in the next few months is to get rid of this implementation versioning macro. These are the current values for this macro.
+
+[frame="none",options="header"]
+|======================
+|Value |Description
+|1     |Initial extension JIT implementation on Intel AMX and DPAS. load, store, mad, fill, piece-wise operations, and the query interface are supported. The exact API used for this implementation is detailed in [matrix extension](doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_deprecated_matrix_no_use.asciidoc)
+|2     |joint matrix type has a new `use` parameter. `layout` on matrix is optional for `C` matrix. JIT implementation on Intel AMX and DPAS. load, store, mad, fill, piece-wise operations, and the query interface are supported 
+|3     |Implementation on Nvidia Tensor Cores
+|======================
+
+## New `joint_matrix` class
+We introduce a new class called `joint_matrix`. The user needs to specify the type of the elements, shape, the matrix use, the memory layout, and the memory scope of the matrix. This results in the following description:
+
+```c++
+namespace sycl::ext::oneapi::experimental::matrix {
+template <typename T, use Use,
+          size_t Rows=sycl::dynamic_extent, size_t Cols=sycl::dynamic_extent,
+          layout Layout = layout::dynamic, typename Group = sub_group>
```

Comment:
> Can you remind me of the rules for using `Layout` if you want portable code that works on both Intel and CUDA? I thought the "A" and "B" matrices couldn't be dynamic if you wanted to run on CUDA, is that right?
> 

Yes the A/B matrices cannot be dynamic for cuda: we use partial specializations for all valid joint_matrix declarations in the cuda backend, so it isn't possible to construct such a use::a/use::b joint_matrix with layout::dynamic in the cuda backend.
> We want the default template parameters to result in portable code, and I see that the template parameter here defaults to `dynamic`. Is that portable?
> 
> Don't we also want a description somewhere in this spec explaining how to use "Layout" if you want portable code? I don't see anything like that now.

FYI for the cuda backend it is also useful for layout::dynamic to be set as default because it means users don't even need to provide a layout template argument when constructing an accumulator joint_matrix: As here in the test for the portable interfaces:

https://github.com/intel/llvm-test-suite/blob/b8a6c135726072ab3e8ab07cb73b2c1d81fbafb0/SYCL/Matrix/joint_matrix_tensorcores.cpp#L151

Reply:
> Yes the A/B matrices cannot be dynamic for cuda: we use partial specializations for all valid joint_matrix declarations in the cuda backend, so it isn't possible to construct such a use::a/use::b joint_matrix with layout::dynamic in the cuda backend.

This restriction isn't conveyed by this spec.  Do we expect this specification to describe a common API that includes CUDA?

> FYI for the cuda backend it is also useful for layout::dynamic to be set as default because it means users don't even need to provide a layout template argument when constructing an accumulator joint_matrix

I agree that this is useful.  This could be achieved by defaulting the `Layout` parameter conditionally, depending on the value of `Use`.

Reply:
Please see under "Layout" subsection:
Important | In both AMX and DPAS support, the layout template parameter can be dynamic for Matrix A and B as well.

I only talk about what is allowed for AMX and DPAS. This is not conveying the restriction about the CUDA backend because I expect Jack to update this file after it is merged to  add the CUDA backend restrictions in joint matrix type (dynamic is the default for only matrix C) and in the joint_matrix_load (load that takes layout as argument is only for C matrix). 

So the restrictions that apply for the CUDA backend will be explained in the specification text by Jack and not imposed in the syntax. If imposed in the syntax, this means we are restricting Intel backend as well.

@JackAKirk, @gmlueck  do you want to proceed differently?

Reply:
> So the restrictions that apply for the CUDA backend will be explained in the specification text by Jack and not imposed in the syntax. If imposed in the syntax, this means we are restricting Intel backend as well.

I thought one of the outcomes of our meeting a few weeks ago was to prioritize the portable case.  Therefore, I think we want the default values of the template parameters to work for both Intel and CUDA.  My concern above is that `Layout` defaults to `dynamic`, but that doesn't work on CUDA for the "A" or "B" matrices.

Reply:
You are right about prioritizing the portable case.  But one of the outcomes of that meeting as well is that the restrictions will still be made in the text of the spec, not in the syntax. Not just yet.

If we make the restriction in the syntax, we are excluding the Intel flexibility.
In the text, we should add that for the CUDA backend, dynamic as default applied to only matrix accumulator.

One of the reasons of that outcome is that I did not want to make more changes at a time. 
The change you are suggesting (in the syntax) should be made in a separate PR after we discuss that specific issue more.

Reply:
Can you please add this as an open issue in the list at the bottom of this spec, then?  I just want to make sure we do not forget about issues like this.

Reply:
Here is a full implementation of the agreed interfaces: https://github.com/intel/llvm/pull/7077 with a working cuda implementation. See that we still use the layout::dynamic default parameter here: https://github.com/intel/llvm/blob/36004a041513635d2567a7d77d9ebab38ba3d29c/sycl/include/sycl/ext/oneapi/matrix/matrix-tensorcores.hpp#L33
but as I mentioned the cuda implementation won't allow a joint_matrix to be constructed with (use::a || use::b) && layout::dynamic.

Reply:
I will add it to open issues.
Note that this is not the only non portable code that is included in this document. We still need to sort things like "packed" layout. 
Once we have the full support of this new API (in a month or so) mentioned by Jack (#7077). I will open up discussion for the non portable scenarios that are still in this document.

Reply:
> Please see under "Layout" subsection: "Important | In both AMX and DPAS support, the layout template parameter can be dynamic for Matrix A and B as well."
> 
> I only talk about what is allowed for AMX and DPAS. This is not conveying the restriction about the CUDA backend because I expect Jack to update this file after it is merged to add the CUDA backend restrictions in joint matrix type (dynamic is the default for only matrix C) and in the joint_matrix_load (load that takes layout as argument is only for C matrix).
> 
> So the restrictions that apply for the CUDA backend will be explained in the specification text by Jack and not imposed in the syntax. If imposed in the syntax, this means we are restricting Intel backend as well.
> 
> @JackAKirk, @gmlueck do you want to proceed differently?

I think this is fine.
If you think that some information is more appropriate in a cuda only extension or intel only extension or the main matrix extension, then I don't mind moving it, especially since these extensions are all experimental. 

I imagine that it is more traditional that extensions are additive rather that subtractive, but I don't think this is a big issue at the moment: so long as the interfaces (note that I don't say "code" here since DPAS and Tensor Cores don't ever have cases where parameters provided to the interfaces are portable) written by a user are fully portable across cuda and intel for the cases where the hardware allows them to (all the practically useful cases) then this is the main thing.